src/main/java/semantics/RDFDatasetDirectStatementDeleter.java

```diff
@@ -0,0 +1,349 @@
+import org.neo4j.graphdb.GraphDatabaseService;
+import org.neo4j.graphdb.Label;
+import org.neo4j.graphdb.Node;
+import org.neo4j.graphdb.Relationship;
+import org.neo4j.graphdb.RelationshipType;
+import org.neo4j.graphdb.Result;
+import org.neo4j.helpers.collection.Iterables;
+import org.neo4j.logging.Log;
+
+/**
+ * This class implements an RDF handler to statement-wise delete imported RDF data sets
+ *
+ * Created on 18/06/2019.
+ *
+ * @author Emre Arkan
+ */
+
+class RDFDatasetDirectStatementDeleter extends RDFDatasetToLPGStatementProcessor implements
+    Callable<Integer> {
+
+  private static final Label RESOURCE = Label.label("Resource");
+
+  private Cache<ContextResource, Node> nodeCache;
+  private long notDeletedStatementCount;
+  private long bNodeCount;
+  private String bNodeInfo;
+
+  RDFDatasetDirectStatementDeleter(GraphDatabaseService db, RDFParserConfig conf, Log l) {
+    super(db, conf);
+    nodeCache = CacheBuilder.newBuilder()
+        .maximumSize(conf.getNodeCacheSize())
+        .build();
+    log = l;
+  }
+
+  @Override
+  public void endRDF() throws RDFHandlerException {
+    Util.inTx(graphdb, this);
+    totalTriplesMapped += mappedTripleCounter;
+    if (parserConfig.getHandleVocabUris() == URL_SHORTEN) {
+      persistNamespaceNode();
+    }
+
+    log.info("Successful (last) partial commit of " + mappedTripleCounter + " triples. " +
+        "Total number of triples deleted is " + totalTriplesMapped + " out of "
+        + totalTriplesParsed + " parsed.");
+  }
+
+  @Override
+  public Integer call() throws Exception {
+
+    for (Map.Entry<ContextResource, Set<String>> entry : resourceLabels.entrySet()) {
+      if (entry.getKey().getUri().startsWith("genid")) {
+        bNodeCount++;
+        notDeletedStatementCount++;
+        continue;
+      }
+      Node tempNode = null;
+      final Node node;
+      try {
+        tempNode = nodeCache.get(entry.getKey(), new Callable<Node>() {
+          @Override
+          public Node call() {
+            Node node = null;
+            Map<String, Object> params = new HashMap<>();
+            String cypher = buildCypher(entry.getKey().getUri(),
+                entry.getKey().getGraphUri(),
+                params);
+            Result result = graphdb.execute(cypher, params);
+            if (result.hasNext()) {
+              node = (Node) result.next().get("n");
+              if (result.hasNext()) {
+                String props =
+                    "{uri: " + entry.getKey().getUri() +
+                        (entry.getKey().getGraphUri() == null ? "}" :
+                            ", graphUri: " + entry.getKey().getGraphUri() + "}");
+                throw new IllegalStateException(
+                    "There are multiple matching nodes for the given properties " + props);
+              }
+            }
+            return node;
+          }
+        });
+      } catch (InvalidCacheLoadException | IllegalStateException e) {
```

Comment:
The method already declares `throws Exception`, so why is this catched?
Also, just printing an error will not stop the rest of the process, which will likely lead to further exceptions.

Therefore the try-catch wrapping should be removed.

Reply:
If the try-catch block is removed the whole deletion process will be interrupted even if only one node cannot be found. To be able to support repetitive deletion, this and the following try-catch blocks have to handle the exceptions thrown.

Reply:
Ok, fair enough.
Maybe we can discuss this tomorrow with Jesus, I think a more elegant solution than `e.printStackStrace()` is required at this point (e.g. a deletion report), but this is probably out-of-scope in this feature branch.